File: libcomponent/src/main/java/com/bear/libcomponent/component/ComponentActivity.java
package com.bear.libcomponent.component;

import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.viewbinding.ViewBinding;

import com.bear.libbase.activity.BaseActivity;

public abstract class ComponentActivity<VB extends ViewBinding> extends BaseActivity<VB> {
    private final ComponentManager componentManager = new ComponentManager();
    
    protected void regActComponent(ActivityComponent<VB> component, Object tag) {
        componentManager.regComponent(this, component, tag);
    }

    protected void regActComponent(ActivityComponent<VB> component) {
        componentManager.regComponent(this, component);
    }

    protected void regComponent(ViewComponent<ViewBinding> component, Object tag) {

    }

    protected void regComponent(ViewComponent<ViewBinding> component) {

    }

    public <C extends GroupComponent> C getComponent(Class<C> clz, Object tag) {
        return componentManager.getComponent(clz, tag);
    }

    public <C extends GroupComponent> C getComponent(Class<C> clz) {
        return componentManager.getComponent(clz);
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        componentManager.dispatchOnBackPressed();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        return componentManager.dispatchOnCreateOptionsMenu(menu, getMenuInflater());
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        return componentManager.dispatchOnOptionsItemSelected(item);
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        componentManager.dispatchOnCreateContextMenu(menu, v, menuInfo);
    }

    @Override
    public boolean onContextItemSelected(@NonNull MenuItem item) {
        return componentManager.dispatchOnContextItemSelected(item);
    }

    public ComponentManager getComponentManager() {
        return componentManager;
    }

    @Override
    protected abstract VB inflateViewBinding(@NonNull LayoutInflater inflater);
}

File: libcomponent/src/main/java/com/bear/libcomponent/host/ComponentActivity.kt
package com.bear.libcomponent.host

import android.view.ContextMenu
import android.view.ContextMenu.ContextMenuInfo
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.viewbinding.ViewBinding
import com.bear.libbase.activity.BaseActivity
import com.bear.libcomponent.core.IComponent
import com.bear.libcomponent.component.ui.ActivityComponent
import com.bear.libcomponent.component.ui.NonUIComponent
import com.bear.libcomponent.component.ui.ViewComponent
import com.bear.libcomponent.core.ComponentManager

// TODO: componentManager 在Activity和Fragment不公用
abstract class ComponentActivity<VB : ViewBinding> : BaseActivity<VB>() {
    internal val componentManager: ComponentManager = ComponentManager()

    override fun onBackPressed() {
        super.onBackPressed()
        componentManager.dispatchOnBackPressed()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        return componentManager.dispatchOnCreateOptionsMenu(menu, menuInflater)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return componentManager.dispatchOnOptionsItemSelected(item)
    }

    override fun onCreateContextMenu(menu: ContextMenu, view: View, menuInfo: ContextMenuInfo) {
        componentManager.dispatchOnCreateContextMenu(menu, view, menuInfo)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        return componentManager.dispatchOnContextItemSelected(item)
    }

    @JvmOverloads
    fun regComponent(component: ActivityComponent<VB>, tag: Any? = null) {
        componentManager.regComponent(this, component, tag)
    }

    @JvmOverloads
    fun regComponent(component: ViewComponent<*>, tag: Any? = null) {
        componentManager.regComponent(this, component, tag)
    }

    @JvmOverloads
    fun regComponent(component: NonUIComponent, tag: Any? = null) {
        componentManager.regComponent(this, component, tag)
    }

    @JvmOverloads
    fun <C : IComponent> getComponent(clz: Class<C>, tag: Any? = null): C? {
        return componentManager.getComponent(clz, tag)
    }

    abstract override fun inflateViewBinding(inflater: LayoutInflater): VB
}
